x-pack/plugins/alerting/server/saved_objects/index.ts

```diff
@@ -129,7 +139,41 @@ export function setupSavedObjects(
         return isRuleExportable(ruleSavedObject, ruleTypeRegistry, logger);
       },
     },
-    modelVersions: ruleModelVersions,
+    modelVersions: {
+      ...ruleModelVersions,
+      '2': encryptedSavedObjects.createModelVersion({
+        modelVersion: {
+          changes: [
+            {
+              /**
+               * I'm not entirely sure about the usage of `unsafe_transform`
+               * here. Probably `data_backfill` is better suited for this data
+               * migration.
+               *
+               * NOTE: We do not introduce schema changes in this migration.
+               * Rule parameters are already typed as `Record<string, any>`,
+               * allowing for any shape of the object. In Security Solution, our
+               * schema is forward compatible; we use zod for runtime type
+               * checking and omit any unknown properties during validation. So
+               * the introduction of new properties in the `params` object looks
+               * safe enough.
+               */
+              type: 'unsafe_transform',
+              transformFn: (document) => {
+                if (isDetectionEngineAADRuleType(document)) {
+                  document.attributes.params.rule_source = {
+                    type: document.attributes.params.immutable ? 'external' : 'internal',
+                  };
+                }
```

Comment:
I think you still need to manually synchronize this new field during save operations, don't you?

Given this scenario:

1. the migration runs, the `params.rule_source` field is populated (but not used)
2. an action causes `params.immutable` to change its value
3. `params.rule_source` will then be out of sync, and broken when the next stage of the upgrade will utilize this field, no?

Reply:
Good point. We'd also want to "hide" rule_source from the HTTP requests and responses during the intermediate release, right?

Reply:
We are going to start writing both fields at the same time before the data migration. I have a draft PR for that already: https://github.com/elastic/kibana/pull/183895.

The plan is:
1. Merge the PR that writes `rule_source` synchronously with `immutable`. That way, all new or modified saved objects will have the new `rule_source` field, but not existing ones. In business logic that reads rules, we will still rely on `immutable`.
2. Release the data migration. Once it's finished, we can be sure that all rule saved objects contain both fields. Our logic will still rely on the `immutable` field.
3. After the migration, we can deprecate the `immutable` field and rely on `rule_source` in all API handlers (except rule import, which will support both formats).

Reply:
> We'd also want to "hide" rule_source from the HTTP requests and responses during the intermediate release, right?

We've already released rule schema changes that define `rule_source` in response schemas as an optional field, so I think we can start returning that field in API responses.